core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -1332,19 +1334,24 @@ public void run() {
             if (serviceConfig != null) {
               effectiveServiceConfig = serviceConfig;
             } else {
-              effectiveServiceConfig = defaultServiceConfig;
+              effectiveServiceConfig = defaultServiceConfig != null
+                  ? defaultServiceConfig : EMPTY_SERVICE_CONFIG;
               if (defaultServiceConfig != null) {
                 channelLogger.log(
                     ChannelLogLevel.INFO,
                     "Received no service config, using default service config");
+              } else {
```

Comment:
Since you need this if-else branch anyway, it would be more readable to integrate the assignment of `effectiveServiceConfig` into this struct, rather than keeping it separate.

Reply:
done.